Oracle.md

```diff
@@ -0,0 +1,106 @@
+
+- start: the first possible outcome number
+- end: the last possible outcome number
+- step: the increment
+
+#### Example: Lottery
+
+An oracle release a random number between 0 and 100 (not a great example sorry...).
+
+```
+start: 0
+end: 100
+step: 1
+```
+
+### Large range
+
+When a range of a numerical outcome is sufficiently large or unbounded, the oracle can represent the outcome using exponent/mantissa (using true normalized form).
+
+The oracle template should include:
+
+- base: the base in which the outcome is represented
+
+The oracle will then release two R value `R_exponent` and `R_mantissa` as well as two signatures `s_exponent` and `s_mantissa`.
+
+#### Example: BTC/USD rate
+
+```
+base: 10
+exponent: 5
+mantissa: 0.9
+```
+
+### Serialization and signature
+
+Oracle templates should be serialized using TLV (to be defined).
+The oracle should provide a signature over the serialized template to the client as an attestation that he committed to release a signature over the event outcome.
+
+## Unique Resource Identifiers
+
+To identify, organize or filter events, it is useful that they are assigned unique resource identifier.
+
+### Categorization
+
+Each segment of the URI path, with the exception of the last one, should represent a (sub)-categorization of the event that is being identified.
+Upon being requested a URI corresponding to a category, the resource provider should return a list of resources included in the category that should correspond to either a sub-category or of an event together with their respective URI (HATEOAS).
+The interpretation of the category semantics is not specified and left to the interpretation of the client (or user).
+
+#### Example: Oracle web server
+
+A client is given the hostname of an oracle web server `https://crystalball.dlc`.
+The client accesses the root resource `/` and is returned a list of event categories served by the oracle with their respective URIs:
+
+- Sport: `/sport`
+- Finance: `/finance`
+- Politics: `/politics`
+
+The client can then request these URIs to access events or sub-categories.
+
+### Events
+
+The last segment of a URI must identify a single event.
+Upon requesting an event URI, a client must be served with:
+
+- The oracle template for the event together with a signature over it
+- The event result and signature(s) over it if available
+
+_NOTE: we could maybe further split the URI into `/signature` and `/template` but that's a detail for now_
+
+An event URI must include a segment guaranteeing its uniqueness, such as a monotonously increasing number or a timestamp.
+
+### Examples:
+
+- Tennis: `/sport/tennis/wimbledon/2019/federer-djokovitch/vs`
+- NBA score: `/sport/nba/s2019/w24/HOU_PHI/vs`
+- NBA score diff: `/sport/nba/s2019/w24/HOU_PHI/diff`
+
+_Note: The URI scheme doesn't have to be bound to http web server and would still be useful in a P2P context such as if using distributed hash tables to request/propagate data for example._
```

Comment:
Not sure if fit for this document, but would be nice to have a way to ask an oracle, "give me every event you support"

Reply:
I think then the oracle could just have a root category `/everything` that contains all the available events if that's what you meant?

Reply:
For me being able to explore the events through the path is enough (at least for the initial spec).